homeassistant/components/vicare/number.py

```diff
@@ -92,23 +101,21 @@ async def async_setup_entry(
     async_add_entities: AddEntitiesCallback,
 ) -> None:
     """Create the ViCare number devices."""
+    entities: list[ViCareNumber] = []
     api = hass.data[DOMAIN][config_entry.entry_id][VICARE_API]
-    circuits = await hass.async_add_executor_job(get_circuits, api)
+    device_config = hass.data[DOMAIN][config_entry.entry_id][VICARE_DEVICE_CONFIG]
 
-    entities: list[ViCareNumber] = []
-    try:
-        for circuit in circuits:
-            for description in CIRCUIT_ENTITY_DESCRIPTIONS:
-                entity = await hass.async_add_executor_job(
-                    _build_entity,
-                    circuit,
-                    hass.data[DOMAIN][config_entry.entry_id][VICARE_DEVICE_CONFIG],
-                    description,
-                )
-                if entity is not None:
-                    entities.append(entity)
-    except PyViCareNotSupportedFeatureError:
-        _LOGGER.debug("No circuits found")
+    circuits = await hass.async_add_executor_job(get_circuits, api)
+    for circuit in circuits:
+        for description in CIRCUIT_ENTITY_DESCRIPTIONS:
+            entity = await hass.async_add_executor_job(
```

Comment:
We should collect all the sync work in a function and schedule that work once on the executor. The context switch from event loop to the executor is expensive.

Reply:
Would it make sense to do that [once in `__init__`](https://github.com/home-assistant/core/blob/b8cc3349be663f55ba1501077cb2d04c0f72540a/homeassistant/components/vicare/__init__.py#L84C5-L101) and store on `hass.data[DOMAIN]`?

Reply:
Entities are a platform concern. It would be ok to collect device information in a central place if it's relevant for more than one entity. But the entities themselves should not be stored anywhere outside their platform.

Reply:
Ah, I misunderstood, I thought this is about `circuits = await hass.async_add_executor_job(get_circuits, api)` which is done for all platforms.

Reply:
I don't understand what the cache has to do with if we make one or multiple sequential executor jobs? It doesn't change how we access the library objects. It only changes how that work is scheduled from the event loop on the executor.

Reply:
Do you think about something like this where the async executor is moved outside the loops?
https://github.com/home-assistant/core/compare/dev...CFenner:sync-4?expand=1

Reply:
Yes, but use a list comprehension. We don't need any local variable for `entity` then.